### PR TITLE
Plans 2023: Woo $1 promotional pricing - trial banner

### PR DESCRIPTION
**Pull request by chriskmnds**
Related to https://github.com/Automattic/martech/issues/2037

## Proposed Changes

Follow up to https://github.com/Automattic/wp-calypso/pull/80430 that introduces the Woo promotional pricing in the Plans grids, to update the wording in the trial banner to mention the promotional pricing info.

- Updates the trial banner shown in both `/my-plan` and `/plans/[ woo-trial ]` (needs confirmation cc @ilyasfoo )
- Some more general refactors. 
    - A `useIntroOffers` hook, which can evolve to grab offers from `/plans` endpoint down the line. This is currently not a React Query hook, although may evolve into one down the line. I couldn't make the call on it right now and had more concerns about it living as a query hook. More granular selectors over the same data will be introduced, so things will evolve, inductively / through use.

### Media

- **Trial banner under `/plans/my-plan/[ woo trial ]`:**

Screenshot 2023-09-06 at 3 15 24 PM

- **Trial banner under `/plans/[ woo trial ]`:**

Screenshot 2023-09-06 at 3 15 39 PM

### General checks

* Sandbox public-api and disable store sandbox (more about this in phab diff D119416-code ) 
    * Go to `/setup/wooexpress` and create a Woo Express site
    * Go to `/plans/[woo-trial]` and confirm pricing shows the intro offers
    * Confirm `/start` renders as previously (no promo pricing)
    * Confirm `/plans/[paid | free site]` renders as previously (no promo pricing)
* Not sandboxing and not proxied (might need to clear browser cache):
    * Confirm `/plans/[woo-trial]` does not show promo offers but original pricing

### Banner

* Sandbox public-api and disable store sandbox (more about this in phab diff D119416-code ) 
    * Confirm `/plans/[ woo trial]` renders the updated banner info
    * Confirm `/plans/my-plan/[ woo trial ]` renders the updated banner info
* Not sandboxing and not proxied (might need to clear browser cache):
    * Confirm `/plans/[ woo trial]` renders the original (trunk) banner info

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by ilyasfoo** [comment]
> Updates the trial banner shown in both /my-plan and /plans/[ woo-trial ] (needs confirmation cc @ilyasfoo )

Good catch, @chriskmnds! Yes this should apply to both UIs

**Comment by chriskmnds** [comment]
@ilyasfoo in progress, but you can start testing already

File: client/my-sites/plans/trials/trial-banner/use-banner-subtitle.ts [hunk 1]
```diff
@@ -1,68 +1,112 @@
...
 	const [ bannerSubtitle, setBannerSubtitle ] = useState( '' );
-
 	// moment.js doesn't have a format option to display the long form in a localized way without the year
 	// https://github.com/moment/moment/issues/3341
 	const readableExpirationDate = trialExpiration?.toDate().toLocaleDateString( locale, {
 		month: 'long',
 		day: 'numeric',
 	} );
+	const introOffers = Plans.useIntroOffers( { siteId: selectedSiteId } );
+	const anyWooExpressIntroOffer = introOffers?.[ PLAN_WOOEXPRESS_MEDIUM ];
```
Comment (ilyasfoo) [inline client/my-sites/plans/trials/trial-banner/use-banner-subtitle.ts]: Nit: I feel this logic might be fragile. Ex: what if in the future medium plan no longer has introductory offer, would it break this logic without us knowing?

Alternatively, we could do a check on multiple plans, adding a unit test, or perhaps build a selector to contain the logic where it's arguably easier to refactor.

File: client/my-sites/plans/trials/trial-banner/use-banner-subtitle.ts [hunk 2]
```diff
@@ -1,68 +1,112 @@
...
 
 	useEffect( () => {
-		let subtitle;
+		// this may need updating for intro offers with singly counted units e.g. 1 month|year
+		const introOfferSubtitle =
+			isEcommerceTrial && anyWooExpressIntroOffer
+				? translate(
+						'Your free trial will end in %(daysLeft)d day. Upgrade by %(expirationdate)s to start selling and take advantage of our limited time offer ' +
+							'-- any Woo Express plan for just %(introOfferFormattedPrice)s a %(introOfferIntervalUnit)s for your first %(introOfferIntervalCount)d %(introOfferIntervalUnit)ss.',
```
Comment (ilyasfoo) [inline client/my-sites/plans/trials/trial-banner/use-banner-subtitle.ts]: Nit: Can we use the unicode character same as design here?
```suggestion
							'— any Woo Express plan for just %(introOfferFormattedPrice)s a %(introOfferIntervalUnit)s for your first %(introOfferIntervalCount)d %(introOfferIntervalUnit)ss.',
```

File: client/my-sites/plans/trials/trial-banner/use-banner-subtitle.ts [hunk 3]
```diff
@@ -1,68 +1,112 @@
...
+						'Your free trial will end in %(daysLeft)d days. Upgrade by %(expirationdate)s to start selling and take advantage of our limited time offer ' +
+							'-- any Woo Express plan for just %(introOfferFormattedPrice)s a %(introOfferIntervalUnit)s for your first %(introOfferIntervalCount)d %(introOfferIntervalUnit)ss.',
```
Comment (ilyasfoo) [inline client/my-sites/plans/trials/trial-banner/use-banner-subtitle.ts]:
```suggestion
							'— any Woo Express plan for just %(introOfferFormattedPrice)s a %(introOfferIntervalUnit)s for your first %(introOfferIntervalCount)d %(introOfferIntervalUnit)ss.',
```

File: packages/data-stores/src/plans/types.ts [hunk 4]
```diff
@@ -105,7 +105,7 @@ export interface PricedAPIPlan {
  * Note: These, unlike the PricedAPIPlan, are returned indexed by product_id (and do not inlcude that in the plan's payload)
  */
 export interface PricedAPISitePlan extends PricedAPIPlanIntroductoryOffer {
-	/* product_id: number; not included in the plan's payload */
+	/* product_id: number; // not included in the plan's payload */
```
Comment (ilyasfoo) [inline packages/data-stores/src/plans/types.ts:108]: Is this change intended?

**Review by ilyasfoo — APPROVED** [review]
This is testing well! I left some nit comments, but pre-approving 🚢 🚢

File: client/my-sites/plans/trials/trial-banner/use-banner-subtitle.ts (hunk 1 above)
Comment (chriskmnds) [reply client/my-sites/plans/trials/trial-banner/use-banner-subtitle.ts]: It's definitely fragile. I feel there are bigger root issues: 

1. how a trial is not bound to any specific plan (so we don't have a reference to look for some offer on it from given endpoint)
2. that we display a generic message "any Woo Express plan for just...", while offers are provided per-plan by said endpoint (so, technically speaking, there can be different offers for different plans). A bit of misalignment between the foundation and use-case I think.

I think the whole thing is fragile by design in that sense.

> Alternatively, we could do a check on multiple plans, adding a unit test, or perhaps build a selector to contain the logic where it's arguably easier to refactor.

I don't know. I can add a loop over multiple plans. Not feeling comfortable engineering more around this though. It feels a little broken by design to me.

File: packages/data-stores/src/plans/types.ts (hunk 4 above)
Comment (chriskmnds) [reply packages/data-stores/src/plans/types.ts:108]: ehm yes and no. think I went a few times back/forth uncommenting during development 😆

File: client/my-sites/plans/trials/trial-banner/use-banner-subtitle.ts (hunk 1 above)
Comment (chriskmnds) [reply client/my-sites/plans/trials/trial-banner/use-banner-subtitle.ts]: @ilyasfoo Thanks for reviewing again.

Made some updates here: 
- introduced a separate getter for only WooExpress plan offers
- then picked off the first from the list to put together the string
- no tests etc.

Can you give this another round of thorough testing pls?

p.s. you might have to ship this yourself if not reviewed by the earliest tomorrow (my time) :-)

Comment (ilyasfoo) [reply client/my-sites/plans/trials/trial-banner/use-banner-subtitle.ts]: > It's definitely fragile. I feel there are bigger root issues:

Good points. I agree that it's not a solid foundation to begin with.

That said, I think the improvements you've made helps. Tested well!

**Review by ilyasfoo — COMMENTED** [review]
Thanks for the improvements, @chriskmnds! This tests well.

However, I'm noticing a [unit test error](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/10767963?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true) that might be related to this PR:

```
(0 , _calypsoConfig.isEnabled) is not a function
TypeError: (0 , _calypsoConfig.isEnabled) is not a function
    at Object.<anonymous> (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/packages/calypso-products/src/plans-list.tsx:3554:15)
    at Runtime._execModule (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-runtime/build/index.js:1430:24)
    at Runtime._loadModule (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-runtime/build/index.js:1013:12)
    at Runtime.requireModule (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-runtime/build/index.js:873:12)
    at Runtime.requireModuleOrMock (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-runtime/build/index.js:1039:21)
    at Object.require (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/packages/calypso-products/src/main.ts:45:1)
    at Runtime._execModule (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-runtime/build/index.js:1430:24)
    at Runtime._loadModule (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-runtime/build/index.js:1013:12)
    at Runtime.requireModule (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-runtime/build/index.js:873:12)
    at Runtime.requireModuleOrMock (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-runtime/build/index.js:1039:21)
    at Object.require (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/packages/calypso-products/src/index.ts:1:1)
    at Runtime._execModule (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-runtime/build/index.js:1430:24)
    at Runtime._loadModule (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-runtime/build/index.js:1013:12)
    at Runtime.requireModule (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-runtime/build/index.js:873:12)
    at Runtime.requireModuleOrMock (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/node_modules/jest-runtime/build/index.js:1039:21)
    at Object.require (/home/teamcity-3/buildAgent/work/c4a9d5b38c1dacde/packages/data-stores/src/plans/hooks/use-intro-offers-for-woo-express.ts:1:1)
...
```

File: client/my-sites/plans/trials/trial-banner/use-banner-subtitle.ts (hunk 1 above)
Comment (chriskmnds) [reply client/my-sites/plans/trials/trial-banner/use-banner-subtitle.ts]: Yeah, it looks more clear and I guess more safe proof right now. Good thinking too.

**Comment by chriskmnds** [comment]
Thanks for reviewing @ilyasfoo . will get a pass and merge soon

> However, I'm noticing a [unit test error](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/10767963?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true) that might be related to this PR:

Indeed, that seemed a little weird. Looks like a mock was missing, which surfaced from the `calypso-products` import 🦆

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8928152

Thank you @chriskmnds for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8928152).
